### PR TITLE
[FEAT][kernels]: add ROCm FlashAttention backend

### PR DESCRIPTION
**Pull request by FED4**
## Summary

Related to #39.

This PR adds ROCm attention support and makes external ROCm FlashAttention the default ROCm attention backend to match the issue contract. PyTorch SDPA remains available as fallback / explicit opt-out:

```bash
export RL_KERNEL_ROCM_ATTN_BACKEND=sdpa
```

The SDPA/AOTriton vs external ROCm FlashAttention performance gap is real, but I am treating that as follow-up work rather than changing the default in this PR.

## Changes

- Add `RocmFlashAttentionOp` for external FlashAttention 2 Triton AMD.
- Use `ROCM_FLASH_ATTN` as the default ROCm `attn` backend.
- Add `NativeAttentionOp` as the PyTorch SDPA fallback for FlashAttention-layout tensors.
- Add `RL_KERNEL_ROCM_ATTN_BACKEND=sdpa` opt-out for PyTorch SDPA dispatch.
- Add GQA/MQA handling to `NativeAttentionOp` for `q_heads > kv_heads`.
- Add CUDA/ROCm/native attention correctness tests against PyTorch SDPA math backend.
- Add ROCm FlashAttention environment checker and installation notes.

## ROCm backend note

This PR does not implement RL-Kernel's generic `ROCM_CK` backend.

For FlashAttention itself, this PR uses FlashAttention 2 Triton AMD. I also tried FlashAttention 2's CK backend, but its standalone build failed on the MI300/gfx942 machine before RL-Kernel was involved, so this PR does not depend on that path.

## Test environment

```text
| Item | Value |
|---|---|
| GPU | AMD Radeon Graphics, gfx942 / MI300-class |
| PyTorch | 2.9.1+rocm6.3 |
| HIP | 6.3.42134-a9a80e791 |
| PyTorch ROCm FA backend | AOTriton |
| External FlashAttention backend | Triton AMD |
```

## Correctness

Command:

```bash
FLASH_ATTENTION_TRITON_AMD_ENABLE=TRUE PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python -m pytest tests/test_attention_correctness.py -q -rs
```

Result:

```text
| Test | Result |
|---|---|
| NativeAttentionOp + ROCm FlashAttention correctness | 87 passed |
| CUDA external FlashAttention cases on ROCm machine | 40 skipped |
| Skip reason | current torch build is not CUDA platform |
```

Observed external ROCm FlashAttention diff vs PyTorch SDPA math backend:

```text
| dtype | max abs diff observed | tolerance |
|---|---:|---:|
| fp16 | ~1.01e-3 | atol=1e-3, rtol=1e-3 |
| bf16 | ~8.84e-3 | atol=2e-2, rtol=2e-2 |
```

Other checks:

```text
| Check | Result |
|---|---|
| python scripts/check_rocm_env.py | passed |
| pytest tests/test_kernel_registry.py rl_engine/tests/test_dispatch.py tests/test_attention_correctness.py -q -rs | 102 passed, 40 skipped |
| ruff check changed files | passed |
| python -m mkdocs build --strict -f mkdocs.yaml | passed |
```

## One-off benchmark

This is a quick sanity benchmark, not a formal performance claim. It compares PyTorch SDPA/AOTriton with external `RocmFlashAttentionOp`.

With `FLASH_ATTENTION_TRITON_AMD_AUTOTUNE=TRUE`:

```text
| dtype | shape (B,S,H,D) | causal | NativeAttentionOp ms | RocmFlashAttentionOp autotune ms | native/external |
|---|---:|---:|---:|---:|---:|
| fp16 | (1,128,4,64) | false | 0.020 | 0.101 | 0.19x |
| fp16 | (2,256,8,128) | false | 0.047 | 0.103 | 0.45x |
| fp16 | (1,512,8,256) | false | 0.099 | 0.116 | 0.85x |
| fp16 | (1,1024,16,64) | false | 0.118 | 0.111 | 1.07x |
| fp16 | (1,2048,16,64) | false | 0.311 | 0.344 | 0.90x |
| bf16 | (1,512,8,256) | false | 0.100 | 0.114 | 0.87x |
| bf16 | (1,1024,16,64) | false | 0.136 | 0.116 | 1.17x |
| bf16 | (1,2048,16,64) | false | 0.345 | 0.377 | 0.92x |
```

External FlashAttention wins on some cases, but not broadly enough to claim a general performance win. Further performance work can investigate CK or shape-specific routing in a separate PR.

## Release Notes

* **New Features**
  * Added ROCm attention backend support, including a configurable attention-kernel selection via `RL_KERNEL_ROCM_ATTN_BACKEND`.
  * Added a PyTorch SDPA-based attention fallback with support for GQA/MQA head layouts.

* **Documentation**
  * Updated ROCm installation guide with build/verification steps and backend switching instructions.

* **Tests**
  * Expanded attention correctness coverage (CUDA/ROCm, causal/non-causal, softmax scaling) and added negative validation cases.
  * Added ROCm dispatch/priority tests for environment-driven kernel selection.

* **Chores**
  * Improved CI token permissions and added an additional targeted pytest run.

### CODE REVIEW

File: rl_engine/kernels/ops/pytorch/attention/__init__.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        k: torch.Tensor,
+        v: torch.Tensor,
+        dropout_p: float = 0.0,
+        softmax_scale: float | None = None,
+        causal: bool = False,
+    ) -> torch.Tensor:
+        q_ref = q.transpose(1, 2)
+        k_ref = k.transpose(1, 2)
+        v_ref = v.transpose(1, 2)
+        out = F.scaled_dot_product_attention(
```
Comment (inaniloquentee) [inline rl_engine/kernels/ops/pytorch/attention/__init__.py:38]: This SDPA call won’t handle GQA/MQA when q_heads > kv_heads, so common LLM shapes may crash. Could we enable GQA or expand k/v, and add a nheads_k < nheads test?

**Comment by FED4** [comment]
@coderabbitai resume

File: scripts/check_rocm_env.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        return hipcc
+
+    fallback = Path("/opt/rocm/bin/hipcc")
+    if fallback.exists():
+        return str(fallback)
+
+    return None
+
+
+def _flash_attn_backend() -> str | None:
```
Comment (Flink-ddd) [inline scripts/check_rocm_env.py]: The CK detection looks inverted. flash_attn_2_cuda is the CUDA-compiled extension; its presence indicates a CUDA build, not a ROCm CK path. Returning "ck" here is semantically wrong on ROCm. The same logic is duplicated in rocm/attention/flash_attn.py and has the same issue. Suggest detecting based on torch.version.hip + actual import success, or just reporting "flash_attn importable: yes/no" instead of guessing the backend.

File: rl_engine/kernels/ops/rocm/attention/flash_attn.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+            self.op = flash_attn_func
+            logger.info("Successfully linked to external flash_attn library (%s backend).", backend)
+        except ImportError as exc:
+            raise RuntimeError(
+                "ROCm FlashAttention requires a ROCm-compatible flash-attn installation. "
+                "See docs/getting_started/installation.md#rocm-backend."
+            ) from exc
+
+    def __call__(
```
Comment (Flink-ddd) [inline rl_engine/kernels/ops/rocm/attention/flash_attn.py:43]: No head_dim validation. The issue asked for explicit supported-head-dim validation and NotImplementedError on unsupported sizes (and this was agreed in the discussion), but __call__ doesn't enforce it. Please add:

```
SUPPORTED_HEAD_DIMS = (64, 128, 256)

if q.shape[-1] not in SUPPORTED_HEAD_DIMS:
    raise NotImplementedError(
        f"RocmFlashAttentionOp supports head_dim in {SUPPORTED_HEAD_DIMS}, "
        f"got {q.shape[-1]}"
    )
```

File: rl_engine/kernels/ops/rocm/attention/flash_attn.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        self,
+        q: torch.Tensor,
+        k: torch.Tensor,
+        v: torch.Tensor,
+        dropout_p: float = 0.0,
+        softmax_scale: float | None = None,
```
Comment (Flink-ddd) [inline rl_engine/kernels/ops/rocm/attention/flash_attn.py:49]: softmax_scale=None is passed straight through and relies on flash_attn_func's internal 1/sqrt(d) default matching the reference's default. The correctness test only passes because both defaults happen to agree; nothing enforces it. Suggest normalizing explicitly so behavior is independent of the downstream library's default:

```
if softmax_scale is None:
    softmax_scale = q.shape[-1] ** -0.5
```

File: rl_engine/kernels/ops/rocm/attention/flash_attn.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        if torch.version.hip is None:
+            raise RuntimeError("RocmFlashAttentionOp requires a ROCm PyTorch build.")
+
+        backend = _select_flash_attn_backend()
+        try:
+            from flash_attn import flash_attn_func
+
+            self.op = flash_attn_func
+            logger.info("Successfully linked to external flash_attn library (%s backend).", backend)
+        except ImportError as exc:
```
Comment (Flink-ddd) [inline rl_engine/kernels/ops/rocm/attention/flash_attn.py]: __init__ only catches ImportError, but the test file's AVAILABILITY_ERRORS is (ImportError, ModuleNotFoundError, OSError, RuntimeError) — flash-attn on ROCm can raise OSError (missing .so) or RuntimeError (HIP runtime) at import. These will leak through here, inconsistent with the test's own expectations.

```
except (ImportError, OSError, RuntimeError) as exc:
    raise RuntimeError(...) from exc
```

File: rl_engine/kernels/ops/rocm/attention/flash_attn.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+import importlib.util
+import os
+
+import torch
+
+from rl_engine.utils.logger import logger
+
+
+def _select_flash_attn_backend() -> str:
```
Comment (Flink-ddd) [inline rl_engine/kernels/ops/rocm/attention/flash_attn.py:13]: This function mutates os.environ["FLASH_ATTENTION_TRITON_AMD_ENABLE"] as a side effect inside what reads like a pure selector. Also note that flash-attn reads this var at import time, so setting it after a prior import won't take effect. Please either set it before any flash_attn import in a documented startup path, or at minimum add a comment that this is an import-time dependency. Same side effect exists in check_rocm_env.py.

File: rl_engine/kernels/registry.py [hunk 7]
```diff
@@ -72,24 +75,43 @@ def __init__(self):
...
                 ],
-                "attn": [OpBackend.FLASH_ATTN, OpBackend.TRITON_GENERIC, OpBackend.PYTORCH_NATIVE],
+                "attn": [OpBackend.FLASH_ATTN, OpBackend.TRITON_GENERIC, OpBackend.PYTORCH_ATTN],
                 "grpo_loss": [OpBackend.TRITON_GRPO_LOSS, OpBackend.PYTORCH_GRPO_LOSS],
                 # Default dispatch logic for new operators
             },
             "rocm": {
                 "logp": [OpBackend.ROCM_AITER, OpBackend.TRITON_GENERIC, OpBackend.PYTORCH_NATIVE],
-                "attn": [OpBackend.TRITON_GENERIC, OpBackend.PYTORCH_NATIVE],
+                "attn": [
```
Comment (Flink-ddd) [inline rl_engine/kernels/registry.py:89]: This makes PYTORCH_ATTN (SDPA) the default ROCm attention backend, with ROCM_FLASH_ATTN demoted to opt-in. Issue #39 asked for ROCm FlashAttention as the primary backend. The benchmark rationale is reasonable, but the PR description itself labels it "not a formal performance claim," and this requirement reversal wasn't agreed in the issue thread. Can you sync this decision back to #39 and get maintainer sign-off before we land the default change? The code is fine; this is a scope/contract question.

File: rl_engine/kernels/registry.py [hunk 8]
```diff
@@ -72,24 +75,43 @@ def __init__(self):
...
-                "attn": [OpBackend.PYTORCH_NATIVE],
+                "attn": [OpBackend.PYTORCH_ATTN],
                 "grpo_loss": [OpBackend.PYTORCH_GRPO_LOSS],
             },
         }
         logger.info(f"KernelRegistry initialized for {device_ctx.device_type}")
+        self._adjust_priority_from_env()
         self._adjust_priority_for_hardware()
 
+    def _adjust_priority_from_env(self):
```
Comment (Flink-ddd) [inline rl_engine/kernels/registry.py:108]: _adjust_priority_from_env() runs before _adjust_priority_for_hardware(). If the hardware pass also touches rocm/attn, an explicit env opt-in could be silently overridden by the hardware default. Explicit env should generally win — please confirm ordering and add a test covering env + hardware together.

File: rl_engine/tests/test_dispatch.py [hunk 9]
```diff
@@ -25,6 +25,23 @@ def test_device_and_registry():
...
+def test_rocm_attention_uses_native_sdpa_by_default(monkeypatch):
+    monkeypatch.delenv("RL_KERNEL_ROCM_ATTN_BACKEND", raising=False)
+
+    registry = KernelRegistry()
+
+    assert registry._priority_map["rocm"]["attn"][0] == OpBackend.PYTORCH_ATTN
+
+
+def test_rocm_attention_flash_attn_opt_in(monkeypatch):
+    monkeypatch.setenv("RL_KERNEL_ROCM_ATTN_BACKEND", " flash_attn ")
```
Comment (Flink-ddd) [inline rl_engine/tests/test_dispatch.py]: Nice catch testing .strip() with " flash_attn ". Missing coverage for the other branches that registry.py actually implements: case-insensitivity, the unknown-value warning path, and explicit "native"/"sdpa". Suggest:

```
@pytest.mark.parametrize("val", ["FLASH_ATTN", "flash-attn", "Flash_Attention"])
def test_rocm_attn_flash_opt_in_case_insensitive(monkeypatch, val):
    monkeypatch.setenv("RL_KERNEL_ROCM_ATTN_BACKEND", val)
    registry = KernelRegistry()
    assert registry._priority_map["rocm"]["attn"][0] == OpBackend.ROCM_FLASH_ATTN
```

File: tests/test_attention_correctness.py [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+        return
+
+    raise RuntimeError("PyTorch SDPA backend selector is unavailable")
+
+
+def should_print_attention_diff():
+    return os.getenv("PRINT_ATTENTION_DIFF", "").lower() in {"1", "true", "yes"}
+
+
+def pytorch_sdpa_reference(q, k, v, *, causal, softmax_scale):
```
Comment (Flink-ddd) [inline tests/test_attention_correctness.py:57]: The reference is computed in fp32 and compared against the low-precision op cast up to float. That validates "low-precision op vs fp32 golden value," which is correct, but it does not validate cross-backend (CUDA vs ROCm) numerical agreement. Suggest noting this in the docstring so future readers don't assume it guarantees backend parity.

**Review by Flink-ddd — CHANGES_REQUESTED** [review]
Thanks for the thorough PR, Happy to re-review once the blocking items are addressed. The core wrapper and test scaffolding are in good shape

**Review by inaniloquentee — CHANGES_REQUESTED** [review]

File: scripts/check_rocm_env.py (hunk 2 above)
Comment (FED4) [reply scripts/check_rocm_env.py]: You're right.  The goal of this script is to check availibility of backends. Guessing backend is a bit redundant and I will make it simpler.

File: rl_engine/kernels/ops/rocm/attention/flash_attn.py (hunk 3 above)
Comment (FED4) [reply rl_engine/kernels/ops/rocm/attention/flash_attn.py:43]: I verified this against the installed ROCm Triton flash-attn backend before adding the guard.

With FLASH_ATTENTION_TRITON_AMD_ENABLE=TRUE, runtime selection reports USE_TRITON_ROCM=True and flash_attn_gpu resolves to flash_attn.flash_attn_triton_amd.interface_fa. I tested fp16/bf16 against the fp32 SDPA reference for causal and non-causal attention. Head dims 32, 40, 59, 64, 65, 80, 96, 111, 112, 127, 128, 129, 160, 192, 224, 256, 257, 384, 511, and 512 pass; 513 fails with Triton shared-memory OutOfResources. 

So I suggest to reject head_dim > 512.

File: rl_engine/kernels/ops/rocm/attention/flash_attn.py (hunk 6 above)
Comment (FED4) [reply rl_engine/kernels/ops/rocm/attention/flash_attn.py:13]: I 'll set it before import.

File: rl_engine/kernels/registry.py (hunk 8 above)
Comment (FED4) [reply rl_engine/kernels/registry.py:108]: You're right, env should be prioritized.

File: rl_engine/kernels/ops/pytorch/attention/__init__.py (hunk 1 above)
Comment (FED4) [reply rl_engine/kernels/ops/pytorch/attention/__init__.py:38]: Since enable_gqa of pytorch SDPA only works on pytorch 2.5+ while we require pytorch 2.4.1, I 'll do expand KV.

File: rl_engine/kernels/registry.py (hunk 7 above)
Comment (FED4) [reply rl_engine/kernels/registry.py:89]: I will still use rocm flash attention as default to keep everything simple in this stage.
Permformance issue can be looked into in a separate PR.

File: rl_engine/kernels/ops/pytorch/attention/__init__.py (hunk 1 above)
Comment (inaniloquentee) [reply rl_engine/kernels/ops/pytorch/attention/__init__.py:38]: > Since enable_gqa of pytorch SDPA only works on pytorch 2.5+ while we require pytorch 2.4.1, I 'll do expand KV.

Makes sense, thanks for keeping PyTorch 2.4.1 compatibility. The manual k/v LGTM.

**Review by inaniloquentee — APPROVED** [review]

**Comment by Flink-ddd** [comment]
@FED4 please resolve CI error first, then We can merge this PR first.

**Comment by Flink-ddd** [comment]
@FED4 please resolve conflicts first, once done, I will merge this PR, Thanks.

**Review by Flink-ddd — APPROVED** [review]
LGTM now, Thank you for your update.
